telnyx_rtc/src/main/java/com/telnyx/webrtc/sdk/TelnyxClient.kt

```diff
@@ -449,25 +456,81 @@ class TelnyxClient(
         Timber.d("ringtone/ringback media player stopped and released")
     }
 
+    private fun requestGatewayStatus() {
+        if (waitingForReg) {
+            socket.send(
+                SendingMessageBody(
+                    id = UUID.randomUUID().toString(),
+                    method = SocketMethod.GATEWAY_STATE.methodName,
+                    params = StateParams(
+                        state = null
+                    )
+                )
+            )
+        }
+    }
 
-    // TxSocketListener Overrides
-    override fun onLoginSuccessful(jsonObject: JsonObject) {
+    /**
+     * Fires once we have successfully received a 'REGED' gateway response, meaning login was successful
+     * @param sessionId, the session ID of the successfully registered session.
+     */
+    internal fun onLoginSuccessful(sessionId: String) {
         Timber.d(
             "[%s] :: onLoginSuccessful [%s]",
             this@TelnyxClient.javaClass.simpleName,
-            jsonObject
+            sessionId
         )
-        sessionId = jsonObject.getAsJsonObject("result").get("sessid").asString
         socketResponseLiveData.postValue(
             SocketResponse.messageReceived(
                 ReceivedMessageBody(
                     SocketMethod.LOGIN.methodName,
-                    LoginResponse(sessionId!!)
+                    LoginResponse(sessionId)
                 )
             )
         )
     }
 
+
+    // TxSocketListener Overrides
+    override fun onClientReady(jsonObject: JsonObject) {
+        Timber.d(
+            "[%s] :: onClientReady :: retrieving gateway state",
+            this@TelnyxClient.javaClass.simpleName,
+        )
+        if (waitingForReg) {
+            requestGatewayStatus()
+            gatewayResponseTimer.schedule(timerTask {
+                if (retryCounter < 2) {
+                    if (waitingForReg) {
+                        onClientReady(jsonObject)
+                    }
+                    retryCounter++
+                } else {
+                    Timber.d(
+                        "[%s] :: Gateway registration has timed out",
+                        this@TelnyxClient.javaClass.simpleName,
+                    )
+                    socketResponseLiveData.postValue(SocketResponse.error("Gateway registration has timed out"))
+                }
+            }, 3000)
```

Comment:
I'm amazed about how many different ways to delay a task over time, we have on Android. Is this blocking UI thread?